main.go

```diff
@@ -103,25 +103,37 @@ func run(l *log.Logger, args arguments) error {
 		l.Println(string(output))
 	}
 
-	var out io.Writer
-	if args.outputFile == "-" {
-		out = os.Stdout
+	var input io.Reader
+	if args.templateFile == "-" {
+		input = os.Stdin
 	} else {
-		var err error
-		out, err = os.Create(args.outputFile)
+		f, err := os.Open(args.templateFile)
 		if err != nil {
 			return err
 		}
+
+		defer f.Close()
+
+		input = f
 	}
 
-	var input io.Reader
-	if args.templateFile == "-" {
-		input = os.Stdin
+	var out io.Writer
+	if args.outputFile == "-" {
+		out = os.Stdout
 	} else {
-		input, err = os.Open(args.templateFile)
+		f, err := os.Create(args.outputFile)
 		if err != nil {
 			return err
 		}
+
+		defer func() {
+			err := f.Close()
+			if err != nil {
```

Comment:
It is better to keep the first error if it is there already. Change to `if finalError == nil {`?

Reply:
`defer encoder.Close()` also returns an error and should be handled the same way. You could write a tiny helper func that takes a pointer to `finalError` and an `io.Closer` to make it all slightly less messy.